src/cockpit/channels/http.py

```diff
@@ -0,0 +1,153 @@
+
+            if 'validate' in opt_tls and not opt_tls['validate']:
+                context.check_hostname = False
+                context.verify_mode = ssl.VerifyMode.CERT_NONE
+
+            connection = http.client.HTTPSConnection(opt_address, opt_port, context=context)
+        else:
+            connection = http.client.HTTPConnection(opt_address, opt_port)
+
+        try:
+            if opt_unix:
+                # create the connection's socket so that it won't call .connect() internally (which only supports TCP)
+                connection.sock = socket.socket(socket.AF_UNIX)
+                connection.sock.connect(opt_unix)
+            else:
+                # explicitly call connect(), so that we can do proper error handling
+                connection.connect()
+        except (OSError, IOError) as e:
+            logger.error('Failed to open %s:%s: %s %s', opt_address, opt_port or opt_unix, type(e), e)
+            problem = 'unknown-hostkey' if isinstance(e, ssl.SSLCertVerificationError) else 'not-found'
+            self.close(problem=problem, message=str(e))
+            return None
+
+        return connection
+
+    def read_send_response(self, response):
+        '''Completely read the response and send it to the channel'''
+
+        while True:
+            # we want to stream data blocks as soon as they come in
+            block = response.read1(4096)
+            if not block:
+                logger.debug('reading response done')
+                # this returns immediately and does not read anything more, but updates the http.client's
+                # internal state machine to "response done"
+                block = response.read()
+                assert block == b''
+                break
+            logger.debug('read block of size %i', len(block))
+            self.loop.call_soon_threadsafe(self.send_data, block)
+
+    def parse_headers(self, http_msg):
+        headers = dict(http_msg)
+        remove = ['Connection', 'Transfer-Encoding']
+        if self.options.get('binary'):
+            remove = ['Content-Length', 'Range']
+        for h in remove:
+            try:
+                del headers[h]
+            except KeyError:
+                pass
+        return headers
+
+    def request(self):
+        connection = self.create_connection()
+        if not connection:
+            # make_connection does the error reporting
+            return
+
+        connection.request(self.options.get('method'),
+                           self.options.get('path'),
+                           headers=self.options.get('headers') or {},
+                           body=self.body)
+        response = connection.getresponse()
+        self.loop.call_soon_threadsafe(lambda: self.send_control(
+            command='response', status=response.status, reason=response.reason,
+            headers=self.parse_headers(response.headers)))
+        self.read_send_response(response)
```

Comment:
Do we need to close the response object or something in order to make sure the connection goes away properly?  Or is that implicit in reading it to EOF?

Reply:
[HTTPResponse](https://docs.python.org/3/library/http.client.html#httpresponse-objects) does not have any kind of .close()/.cleanup() etc. In Python I'd expect the `connection` object to close itself when it goes out of scope, but I added an explicit `connection.close()` now just to be on the safe side.